clang/test/SemaSYCL/zero-length-arrays.cpp

```diff
@@ -0,0 +1,162 @@
+      (void)Var; // expected-error 2{{zero-length arrays are not permitted in SYCL device code}}
+    });
+  });
+  // expected-error@#KernelSingleTaskKernelFuncCall {{zero-length arrays are not permitted in SYCL device code}}
+  // expected-note@#KernelSingleTask {{in instantiation of function template specialization}}
+  q.submit([&](handler &h) {
+    // expected-note@+2 {{in instantiation of function template specialization}}
+    // expected-note@+1 {{within field of type 'Templated<0U, float>' declared here}}
+    h.single_task<class TempContext1>([Var] {
+    });
+  });
+}
+
+void foo(const unsigned X) {
+  int Arr[0]; // expected-note 2{{declared here}}
+  ZEROARR TypeDef; // expected-note {{declared here}}
+  ZEROARR *Ptr; // expected-note {{declared here}}
+  // expected-note@#KernelSingleTaskKernelFuncCall {{called by 'kernel_single_task<Simple, (lambda at}}
+  // expected-error@#KernelSingleTaskKernelFuncCall 3{{zero-length arrays are not permitted in SYCL device code}}
+  // expected-note@#KernelSingleTask {{in instantiation of function template specialization}}
+  q.submit([&](handler &h) {
+    // expected-note@+1 {{in instantiation of function template specialization}}
+    h.single_task<class Simple>([=]() {
+      (void)Arr; // expected-error {{zero-length arrays are not permitted in SYCL device code}}
+      (void)TypeDef; // expected-error {{zero-length arrays are not permitted in SYCL device code}}
+      // expected-note@+1 {{field of illegal pointer type 'ZEROARR *' (aka 'float (*)[0]') declared here}}
+      (void)Ptr; // expected-error {{zero-length arrays are not permitted in SYCL device code}}
+    });
+  });
+  // expected-note@#KernelSingleTaskKernelFuncCall {{called by 'kernel_single_task<Simple1, (lambda at}}
+  // expected-error@#KernelSingleTaskKernelFuncCall {{zero-length arrays are not permitted in SYCL device code}}
+  // expected-note@#KernelSingleTask {{in instantiation of function template specialization}}
+  q.submit([&](handler &h) {
+    // expected-note@+2 {{in instantiation of function template specialization}}
+    // expected-note@+1 {{field of illegal type 'int[0]' declared here}}
+    h.single_task<class Simple1>([Arr]{ // expected-error {{zero-length arrays are not permitted in SYCL device code}}
+    });
+  });
+  WrapperOfWrapper St;
+  // expected-note@#KernelSingleTaskKernelFuncCall {{called by 'kernel_single_task<SimpleStruct, (lambda at}}
+  // expected-error@#KernelSingleTaskKernelFuncCall 2{{zero-length arrays are not permitted in SYCL device code}}
+  // expected-note@#KernelSingleTask {{in instantiation of function template specialization}}
+  q.submit([&](handler &h) {
+    // expected-note@+1 {{in instantiation of function template specialization}}
+    h.single_task<class SimpleStruct>([=]{
+      // expected-note@+1 {{within field of type 'WrapperOfWrapper' declared here}}
+      (void)St.F.BadArray; // expected-error 4{{zero-length arrays are not permitted in SYCL device code}}
```

Comment:
Why is this diagnostic generated 4 times for this declaration?

Reply:
Two errors for `WrapperOfWrapper` struct because it contains two arrays, one for use of field `F` because it has an array inside, another one for use of `BadArray` because it is array.

Reply:
Oh right. Thanks for explaining!